R/nhdplusTools.R

```diff
@@ -1,25 +1,80 @@
+# NHDPlus Attributes
+COMID <- FEATUREID <-
+  Hydroseq <- DnHydroseq <- DnMinorHyd <- LevelPathI <-
+  ToNode <- FromNode <-
+  TotDASqKM <- LENGTHKM <-
+  Pathlength <- StreamCalc <- StreamOrde <- TerminalFl <-
+  Divergence <- TerminalPa <- StartFlag <- FTYPE <-
+  FromMeas <- ToMeas <- REACHCODE <- REACH_meas <- NULL
+
+  # Package Attribute Names
+COMID.y <- ID <- becomes <- ds_num_upstream <- fID <-
+  dsLENGTHKM <- ds_joined_fromCOMID <- fromCOMID <-
+  fromLENGTHKM <- fromTotDASqKM <- geom_len <-
   geometry <- join_category <- joined_fromCOMID <-
   joined_fromCOMID_new <- joined_toCOMID <- member_COMID <-
   new_joined_fromCOMID <- new_joined_toCOMID <- new_toCOMID <-
   num_upstream <- part <- piece <- pieces <- removed_COMID <-
   split_fID <- toCOMID <- toID <- usLENGTHKM <- usTotDASqKM <-
   . <- L1 <- X <- Y <- breaks <- dist_ratio <- ideal_len <-
   len <- nID <- new_index <- piece_len <- setNames <- start <-
-  FromMeas <- REACHCODE <- REACH_meas <- ToMeas <-
-  index <- measure <- nn.idx <- precision_index <- FEATUREID <- NULL
+  index <- measure <- nn.idx <- precision_index <- NULL
 
 nhdplusTools_env <- new.env()
 
 default_nhdplus_path <- "../NHDPlusV21_National_Seamless.gdb"
 
+assign("prepare_nhdplus_attributes",
+       c("COMID", "LENGTHKM", "FTYPE", "TerminalFl",
+         "FromNode", "ToNode", "TotDASqKM",
+         "StartFlag", "StreamOrde", "StreamCalc",
+         "TerminalPa", "Pathlength", "Divergence"),
+       envir = nhdplusTools_env)
+
+assign("split_flowlines_attributes",
+       c("COMID", "toCOMID", "LENGTHKM", "TotDASqKM"),
+       envir = nhdplusTools_env)
+
+assign("collapse_flowlines_attributes",
+       c("COMID", "toCOMID", "LENGTHKM", "TotDASqKM"),
+       envir = nhdplusTools_env)
+
+assign("reconcile_collapsed_flowlines_attributes",
+       c("COMID", "toCOMID", "LENGTHKM", "TotDASqKM"),
+       envir = nhdplusTools_env)
+
+assign("get_UT_attributes",
+       c("COMID", "Pathlength", "LENGTHKM", "Hydroseq"),
+       envir = nhdplusTools_env)
+
+assign("get_UM_attributes",
+       c("COMID", "Pathlength", "LevelPathI",
+         "UpHydroseq", "Hydroseq"),
+       envir = nhdplusTools_env)
+
+assign("get_DM_attributes",
+       c("COMID", "LENGTHKM", "DnHydroseq", "Hydroseq"),
+       envir = nhdplusTools_env)
+
+assign("get_DD_attributes",
+       c("COMID", "DnMinorHyd", "DnHydroseq", "Hydroseq"),
+       envir = nhdplusTools_env)
+
+assign("get_flowline_index_attributes",
+       c("COMID", "REACHCODE", "ToMeas", "FromMeas"),
+       envir = nhdplusTools_env)
+
+
+check_names <- function(names_flines, function_name) {
+  expect_names <- get(paste0(function_name, "_attributes"),
+                      envir = nhdplusTools_env)
+  if ( !all(expect_names %in% names_flines)) {
+    stop(paste0("Missing some required attributes in call to: ",
+                function_name, ". Expected all of: ",
+                paste(expect_names, collapse = ", "), "."))
```

Comment:
Maybe instead:
```
stop(paste0("Missing some required attributes in call to: ",
 +                function_name, ". Expected: ",
 +                paste(expect_names[which(!(expect_names %in% names_flines))], collapse = ", "), "."))
```
So they don't even have to figure out what they're missing.

Reply:
👍 You are too nice.